### PR TITLE
modularizes attachments code.

### PR DESCRIPTION
**Pull request by SplinterGP**
## About The Pull Request
Modularizes attachments on guns so they use an attachments assoc lazylist, so you can add new attachments slots just by giving it to an attachment, without needing a specific var for it on guns.

## Changelog
:cl:
refactor: refactors how the guns track what attachments it has, doens't affect gameplay.
/:cl:

### CODE REVIEW

File: code/modules/projectiles/gun_helpers.dm [hunk 1]
```diff
@@ -312,23 +310,9 @@ should be alright.
...
-			if(under && !(under.flags_attach_features & ATTACH_REMOVABLE))
-				can_attach = FALSE
-		if(ATTACHMENT_SLOT_STOCK)
-			if(stock && !(stock.flags_attach_features & ATTACH_REMOVABLE))
-				can_attach = FALSE
-
-	if(!can_attach)
+	//Checks if there is any unremovable attachment on our slot, if true, return.
+	var/obj/item/attachable/currently_in_slot = LAZYACCESS(attachments, attachment.slot)
+	if(currently_in_slot && !(currently_in_slot.flags_attach_features & ATTACH_REMOVABLE))
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_helpers.dm:315]: why not ?.

File: code/modules/projectiles/gun_system.dm [hunk 2]
```diff
@@ -95,10 +95,7 @@
 	var/list/attachable_overlays	= list(ATTACHMENT_SLOT_MUZZLE, ATTACHMENT_SLOT_RAIL, ATTACHMENT_SLOT_UNDER, ATTACHMENT_SLOT_STOCK, ATTACHMENT_SLOT_MAGAZINE) //List of overlays so we can switch them in an out, instead of using Cut() on overlays.
 	var/list/attachable_offset 		= null		//Is a list, see examples of from the other files. Initiated on New() because lists don't initial() properly.
 	var/list/attachable_allowed		= null		//Must be the exact path to the attachment present in the list. Empty list for a default.
-	var/obj/item/attachable/muzzle 	= null		//Attachable slots. Only one item per slot.
-	var/obj/item/attachable/rail 	= null
-	var/obj/item/attachable/under 	= null
-	var/obj/item/attachable/stock 	= null
+	var/list/obj/item/attachable/attachments
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_system.dm:98]: Autodoc

File: code/modules/projectiles/guns/specialist.dm [hunk 3]
```diff
@@ -120,7 +120,8 @@
 	. = ..()
 
 /obj/item/weapon/gun/rifle/sniper/antimaterial/process()
-	if(!rail.zoom)
+	var/obj/item/attachable/scope = LAZYACCESS(attachments, ATTACHMENT_SLOT_RAIL)
+	if(scope && !scope.zoom)
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/specialist.dm:124]: didnt null check before why does it now

File: code/modules/projectiles/guns/specialist.dm [hunk 4]
```diff
@@ -138,7 +139,8 @@
 
 /obj/item/weapon/gun/rifle/sniper/antimaterial/zoom(mob/living/user, tileoffset = 11, viewsize = 12) //tileoffset is client view offset in the direction the user is facing. viewsize is how far out this thing zooms. 7 is normal view
 	. = ..()
-	if(!rail.zoom && (targetmarker_on || targetmarker_primed) )
+	var/obj/item/attachable/scope = LAZYACCESS(attachments, ATTACHMENT_SLOT_RAIL)
+	if((scope && !scope.zoom) && (targetmarker_on || targetmarker_primed) )
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/specialist.dm:143]: ditto

File: code/modules/projectiles/guns/specialist.dm [hunk 5]
```diff
@@ -178,7 +180,8 @@
 
 
 /obj/item/weapon/gun/rifle/sniper/antimaterial/proc/laser_on(mob/user)
-	if(!rail.zoom) //Can only use and prime the laser targeter when zoomed.
+	var/obj/item/attachable/scope = LAZYACCESS(attachments, ATTACHMENT_SLOT_RAIL)
+	if(scope && !scope.zoom) //Can only use and prime the laser targeter when zoomed.
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/specialist.dm:184]: ditto

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_helpers.dm (hunk 1 above)
Comment (SplinterGP) [reply code/modules/projectiles/gun_helpers.dm:315]: i banana myself when doing ! and ?.

Comment (SplinterGP) [reply code/modules/projectiles/gun_helpers.dm:315]: if(null && !(null.var & flag)) returns false

if(!null?.var & flag) returns true

Comment (TiviPlus) [reply code/modules/projectiles/gun_helpers.dm:315]: That's not what I meant but whatever

File: code/modules/projectiles/gun_system.dm (hunk 2 above)
Comment (SplinterGP) [inline code/modules/projectiles/gun_system.dm:98]:
```suggestion
	///assoc list of "slot" = ref for attachments, supports any slot, two attachments of same slot cant exist.
	var/list/obj/item/attachable/attachments
```

File: code/modules/projectiles/guns/specialist.dm (hunk 5 above)
Comment (SplinterGP) [inline code/modules/projectiles/guns/specialist.dm:184]:
```suggestion
	if(!scope.zoom) //Can only use and prime the laser targeter when zoomed.
```

File: code/modules/projectiles/gun_system.dm [hunk 6]
```diff
@@ -177,14 +175,11 @@
...
-		QDEL_NULL(rail)
-	if(under)
-		QDEL_NULL(under)
-	if(stock)
-		QDEL_NULL(stock)
+
+	if(attachments)
+		for(var/slot in attachments)
+			qdel(attachments[slot])
+			LAZYREMOVE(attachments, slot)
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_system.dm:182]: just unsetempty at the end of this loop

File: code/modules/projectiles/guns/specialist.dm (hunk 3 above)
Comment (TiviPlus) [reply code/modules/projectiles/guns/specialist.dm:124]: standing

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**